src/RadixDlt.NetworkGateway.Abstractions/CoreCommunications/TransactionSubmitter.cs

```diff
@@ -0,0 +1,240 @@
+ * permissions under this License.
+ */
+
+using RadixDlt.CoreApiSdk.Api;
+using RadixDlt.CoreApiSdk.Model;
+using RadixDlt.NetworkGateway.Abstractions.Extensions;
+using RadixDlt.NetworkGateway.Abstractions.Model;
+using System;
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace RadixDlt.NetworkGateway.Abstractions.CoreCommunications;
+
+public class TransactionSubmitter
+{
+    public static async Task<NodeSubmissionResult> Submit(
+        TransactionApi transactionApi,
+        string networkName,
+        byte[] notarizedTransaction,
+        SubmitContext submitContext,
+        IReadOnlyCollection<ITransactionSubmitterObserver> observers,
+        TimeSpan submissionTimeout,
+        CancellationToken cancellationToken)
+    {
+        await observers.ForEachAsync(observer => observer.ObserveSubmitAttempt(submitContext));
+
+        using var ctsWithSubmissionTimeout = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken);
+        ctsWithSubmissionTimeout.CancelAfter(submissionTimeout);
+
+        var result = await SubmitInternal(transactionApi, networkName, notarizedTransaction, ctsWithSubmissionTimeout.Token);
+
+        await observers.ForEachAsync(observer => observer.ObserveSubmitResult(submitContext, result));
+
+        return result;
+    }
+
+    private static async Task<NodeSubmissionResult> SubmitInternal(
+        TransactionApi transactionApi,
+        string networkName,
+        byte[] notarizedTransaction,
+        CancellationToken cancellationToken)
+    {
+        try
+        {
+            var result = await CoreApiErrorWrapper.ResultOrError<TransactionSubmitResponse, TransactionSubmitErrorResponse>(() =>
+                transactionApi.TransactionSubmitPostAsync(
+                    new TransactionSubmitRequest(
+                        network: networkName,
+                        notarizedTransactionHex: notarizedTransaction.ToHex()
+                    ),
+                    cancellationToken
+                ));
+
+            if (result.Succeeded)
+            {
+                return new NodeSubmissionResult.AcceptedIntoMempool(IsAlreadyThere: result.SuccessResponse.Duplicate);
+            }
+
+            return result.FailureResponse.Details switch
+            {
+                TransactionSubmitIntentAlreadyCommitted alreadyCommittedDetails =>
+                    new NodeSubmissionResult.IntentAlreadyCommitted(alreadyCommittedDetails.CommittedAs),
+                TransactionSubmitPriorityThresholdNotMetErrorDetails priorityThresholdDetails =>
+                    new NodeSubmissionResult.MempoolPriorityThresholdNotMet(priorityThresholdDetails),
+                TransactionSubmitRejectedErrorDetails rejectedDetails =>
+                    new NodeSubmissionResult.Rejection(rejectedDetails),
+                _ => throw new ArgumentOutOfRangeException(nameof(result.FailureResponse.Details)),
```

Comment:
Please use at least `new ArgumentOutOfRangeException(nameof(...), details, null)` - otherwise we won't know which unexpected/unhandled type was actually observed.